### PR TITLE
Add smoke-dataset evaluation part

### PR DESCRIPTION
**Pull request by zurk**
Work in progress!
This PR depends on a bunch of previous ones.

For now, there is only approximate architecture. All code needs to be polished.

### CODE REVIEW

File: lookout/core/tests/server.py [hunk 1]
```diff
@@ -65,10 +65,9 @@ def run(cmd: str, fr: str, to: str, port: int, git_dir: str, config_json: str=No
         "--git-dir", git_dir,
     ]
     if config_json:
-        command.extend(("--config_json", config_json))
-    subprocess.run(
-        command,
-        stdout=sys.stdout, stderr=sys.stderr, check=True)
+        command.extend(("--config-json", config_json))
+    subprocess.run(command,
```
Comment (vmarkovtsev) [inline lookout/core/tests/server.py]: Change not needed

**Comment by zurk** [comment]
There are still a lot of places to improve this code like docstrings and `generate_file` function partially implemented, but I think it is ready for first review round.

**Comment by vmarkovtsev** [comment]
@zurk Can you please rebase now

**Comment by zurk** [comment]
done

File: lookout/style/format/analyzer.py [hunk 2]
```diff
@@ -138,6 +140,21 @@ def group_line_nodes(y, y_pred, vnodes_y, rule_winners):
                     comments.append(comment)
         return comments
 
+    @staticmethod
+    def generate_file(vnodes_y, y_pred, vnodes) -> List[str]:
+        """
+        TODO (zurk).
+        """
+        indentations = {CLS_SPACE_INC, CLS_TAB_INC, CLS_SPACE_DEC, CLS_TAB_DEC}
+        indentations = set(CLASS_INDEX[x] for x in indentations)
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]:
```python
indentations = {CLASS_INDEX[x] for x in [CLS_SPACE_INC, CLS_TAB_INC, CLS_SPACE_DEC, CLS_TAB_DEC]}
```

File: lookout/style/format/analyzer.py [hunk 3]
```diff
@@ -138,6 +140,21 @@ def group_line_nodes(y, y_pred, vnodes_y, rule_winners):
...
+        for vnode, y_ in zip(vnodes_y, y_pred):
+            if vnode.y != y_:
+                if vnode.y in indentations or y_ in indentations:
+                    raise NotImplementedError()
+                vnode.y = y_
+                vnode.value = CLS_TO_STR[CLASSES[y_]]
+        return "".join([vnode.value for vnode in vnodes]).splitlines()
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: `[]` is not needed

File: lookout/style/format/benchmarks/evaluate_smoke.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        matcher.ratio()
+        for action, _, _, _, _ in matcher.get_opcodes():
+            if action == "equal":
+                continue
+            if action == "delete":
+                err += 1
+            elif action == "insert":
+                err += 1
+            elif action == "replace":
+                err += 2
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/evaluate_smoke.py]: Why 2?

File: lookout/style/format/benchmarks/evaluate_smoke.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                server.run("push", fr=row["from"], to=row["to"], port=port,
+                           git_dir=str(repopath), )
+                server.run("review", fr=row["from"], to=row["to"], port=port,
+                           git_dir=str(repopath),
+                           config_json=json.dumps(config_json))
+                pass
+        log.info("Quality report saved to %s", reportpath)
+    finally:
+        fs.cleanup()
+        context_manager.__exit__(None, None, None)
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/evaluate_smoke.py]: why not `with context_manager:`? It should perfectly fit

File: lookout/style/format/benchmarks/evaluate_smoke.py (hunk 4 above)
Comment (zurk) [reply lookout/style/format/benchmarks/evaluate_smoke.py]: because it is one delete and one incert

File: lookout/style/format/analyzer.py (hunk 3 above)
Comment (zurk) [reply lookout/style/format/analyzer.py]:
```suggestion
return "".join(vnode.value for vnode in vnodes).splitlines()
```

File: lookout/style/format/benchmarks/evaluate_smoke.py (hunk 5 above)
Comment (zurk) [reply lookout/style/format/benchmarks/evaluate_smoke.py]: right!

File: lookout/style/format/benchmarks/evaluate_smoke.py (hunk 4 above)
Comment (vmarkovtsev) [reply lookout/style/format/benchmarks/evaluate_smoke.py]: Imagine that a model predicts a wrong token. It is a single mistake, not a double one.

File: lookout/style/format/benchmarks/evaluate_smoke.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+                            "style_name": row["style"],
+                            "report_path": reportpath
+                        }
+                    }
+                    server.run("push", fr=row["from"], to=row["to"], port=port,
+                               git_dir=str(repopath), )
+                    server.run("review", fr=row["from"], to=row["to"], port=port,
+                               git_dir=str(repopath),
+                               config_json=json.dumps(config_json))
+                    break
```
[the remark this answers is not in the document]
Comment (zurk) [reply lookout/style/format/benchmarks/evaluate_smoke.py]: oops. debug code.

File: lookout/style/format/benchmarks/evaluate_smoke.py (hunk 4 above)
Comment (zurk) [reply lookout/style/format/benchmarks/evaluate_smoke.py]: ok

**Comment by vmarkovtsev** [comment]
@zurk Ping :)

File: lookout/style/format/analyzer.py [hunk 7]
```diff
@@ -142,6 +144,21 @@ def group_line_nodes(y, y_pred, vnodes_y, rule_winners):
                     comments.append(comment)
         return comments
 
+    @staticmethod
+    def generate_file(vnodes_y, y_pred, vnodes) -> List[str]:
+        """
+        TODO (zurk).
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: todo

**Pull request merged**